qiskit/transpiler/target.py

```diff
@@ -0,0 +1,600 @@
+    the target like::
+
+        from qiskit.transpiler import Target, InstructionProperties
+        from qiskit.circuit.library import UGate, CXGate
+        from qiskit.circuit import Parameter
+
+        gmap = Target()
+        theta = Parameter('theta')
+        phi = Parameter('phi')
+        lam = Parameter('lambda')
+        u_props = {
+            (0,): InstructionProperties(length=5.23e-8, error=0.00038115),
+            (1,): InstructionProperties(length=4.52e-8, error=0.00032115),
+        }
+        gmap.add_instruction(UGate(theta, phi, lam), u_props)
+        cx_props = {
+            (0,1): InstructionProperties(length=5.23e-7, error=0.00098115),
+            (1,0): InstructionProperties(length=4.52e-7, error=0.00132115),
+        }
+        gmap.add_instruction(CXGate(), cx_props)
+
+    The intent of the ``Target`` object is to inform Qiskit's compiler about
+    the constraints of a particular backend so the compiler can compile an
+    input circuit to something that works and is optimized for a device.
+
+    Each instruction in the Target is indexed by a unique string name that uniquely
+    identifies that instance of an :class:`~qiskit.circuit.Instruction` object in
+    the Target. There is a 1:1 mapping between a name and an
+    :class:`~qiskit.circuit.Instruction` instance in the target and each name must
+    be unique. By default the name is the :attr:`~qiskit.circuit.Instruction.name`
+    attribute of the instruction, but can be set to anything. This lets a single
+    target have multiple instances of the same instruction class with different
+    parameters. For example, if a backend target has two instances of an
+    :class:`~qiskit.circuit.library.RXGate` one is parameterized over any theta
+    while the other is tuned up for a theta of pi/6 you can add these by doing something
+    like::
+
+        import math
+
+        from qiskit.transpiler import Target, InstructionProperties
+        from qiskit.circuit.library import RXGate
+        from qiskit.circuit import Parameter
+
+        target = Target()
+        theta = Parameter('theta')
+        rx_props = {
+            (0,): InstructionProperties(length=5.23e-8, error=0.00038115),
+        }
+        target.add_instruction(RXGate(theta), rx_props)
+        rx_30_props = {
+            (0,): InstructionProperties(length=5.23e-6, error=1.2e-5
+        }
+        target.add_instruction(RXGate(math.pi / 6), rx_30_props, name='rx_30')
+
+    Then in the ``target`` object accessing by ``rx_30`` will get the fixed
+    angle :class:`~qiskit.circuit.library.RXGate` while ``rx`` will get the
+    parameterized :class:`~qiskit.circuit.library.RXGate`.
+
+    .. note::
+
+        This class assumes that qubit indices start at 0 and are a contiguous
+        set if you want a submapping the bits will need to be reindexed in
+        a new object
+
+    .. note::
+
+        This class is designed to be additive only for gates, qargs, and qubits.
```

Comment:
Here I just meant qargs as in the list of [(0,) (1, 0), (1,)] vs qubits being the equivalent to `list(range(max(qargs)))`. Its me probably being to concerned with the internal implementation details because for the internal caching we store the max qubit value on the fly as we add new instructions and track the qubit list based on that, but reasoning about removals and this list is hard (qargs is a bit easier as it's explicitly defined everywhere).

Reply:
```suggestion
        This class only supports additions of gates, qargs, and qubits.
```